packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -405,6 +430,108 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
+        address, /*_initiator*/
+        address _asset,
+        uint256 _amount,
+        uint256 _premium,
+        uint8 _callSource,
+        bytes memory _calldata
+    ) internal override {
+        if (CALLBACK_SOURCE(_callSource) == CALLBACK_SOURCE.FLASHLOAN_W_MINT_DEPOSIT_NFT) {
+            console.log("_amount", _amount);
+            console.log("_premium", _premium);
+            console.log(
+                "IWETH9(weth).balanceOf(address(this) when receiving flashloan",
+                IWETH9(weth).balanceOf(address(this))
+            );
+
+            FlashloanMintDepositNftData memory data = abi.decode(_calldata, (FlashloanMintDepositNftData));
+
+            // convert flashloaned WETH to ETH
+            IWETH9(weth).withdraw(_amount);
+
+            uint256 vaultId = IController(controller).mintWPowerPerpAmount{value: data.collateralToMint}(
+                data.vaultId,
+                data.wPowerPerpAmount,
+                0
+            );
+
+            // LP data.wPowerPerpAmount & data.collateralToLp in Uni v3
+            uint256 amount0Desired = isWethToken0 ? data.collateralToLp : data.wPowerPerpAmount;
+            uint256 amount1Desired = isWethToken0 ? data.wPowerPerpAmount : data.collateralToLp;
+            uint256 uniTokenId = _lpWPowerPerpPool(
+                address(this),
+                data.collateralToLp,
+                amount0Desired,
+                amount1Desired,
+                data.lpAmount0Min,
+                data.lpAmount1Min,
+                block.timestamp,
+                int24(data.lpLowerTick),
+                int24(data.lpUpperTick)
+            );
+            console.log("uniTokenId", uniTokenId);
+
+            // deposit Uni NFT token in vault
+            INonfungiblePositionManager(nonfungiblePositionManager).approve(controller, uniTokenId);
+            IController(controller).mintWPowerPerpAmount(vaultId, 0, uniTokenId);
+
+            console.log("address(this).balance just before withdrawing ETH", address(this).balance);
+            console.log(
+                "IWETH9(weth).balanceOf(address(this) just before withdrawing ETH",
+                IWETH9(weth).balanceOf(address(this))
+            );
+
+            // remove flashloan amount in ETH from vault + any amount of collateral user want to withdraw (sum <= vault.collateralAmount)
+            IController(controller).burnWPowerPerpAmount(vaultId, 0, _amount.add(data.collateralToWithdraw));
+
+            console.log("address(this).balance just after withdrawing ETH", address(this).balance);
+            console.log(
+                "IWETH9(weth).balanceOf(address(this) just after withdrawing ETH",
+                IWETH9(weth).balanceOf(address(this))
+            );
+            console.log("_amount.add(_premium)", _amount.add(_premium));
+
+            // convert flashloaned amount + fee from ETH to WETH to prepare for payback
+            IWETH9(weth).deposit{value: _amount.add(_premium)}();
```

Comment:
need to return ETH to user and any unused oSQTH (say if not 100% of the desired oSQTH was able to be LP'd with)